releases-implementation/README.md

```diff
@@ -0,0 +1,415 @@
+
+## 4. Proposal
+
+This section defines how releases are implemented, where they are, how they are developed, tested and delivered.
+
+Then it compares the proposed solution to the current one that we have for our new KaaS product that is based on Cluster API.
+
+### 4.1. Implementation of releases
+
+This document proposes to continue using `giantswarm/releases` repository for creating and delivering releases, albeit in a simplified way when compared to the vintage releases. Release resources are used in a minimal way, and cluster-$provider apps continue to be used for deploying workload clusters.
+
+Briefly put, cluster-$provider apps are still deployed in almost exactly same way, with the following few differences:
+- In cluster-$provider app manifest, instead of specifying `.spec.version`, we MUST specify `release.giantswarm.io/version` label.
+- Information about app version, catalog and dependencies MUST BE obtained from the Release resource (during Helm rendering phase) when deploying a production cluster.
+- Information about app version, catalog and dependencies CAN be overridden with Helm values when deploying a test cluster in e2e tests.
+
+The core of this proposal is the idea to decouple the app and component versions from the cluster-$provider apps, which will then enable us to have a very scalable process for working with releases, where we can easily create and manage many releases across many providers and develop mechanisms to enforce business logic across all of those.
+
+We will start with showing a slightly simplified version of the Release resource, then explain how it is delivered and finally consumed.
+
+Other parts the `giantswarm/releases` repository that we CAN continue to use are:
+- release notes,
+- announcements,
+- unit tests for releases (for enforcing various rules about releases).
+
+`requests.yaml` file usage SHOULD be deprecated in favour of more scalable and automated process based on GitHub actions and Renovate.
+
+#### 4.1.1. Creating new release
+
+With the current Release CRD definition, we create a Release resource in the following way:
+- All apps, both those deployed as App resources and as HelmRelease resources, MUST be added to Release `.spec.apps`. For every app:
+	- we MUST specify app name,
+	- we MUST specify app version,
+	- we CAN OPTIONALLY specify the name of the catalog from which the app is installed (“default” by default), and
+	- we CAN OPTIONALLY specify the app’s dependencies.
+- Kubernetes version MUST BE specified as a `.spec.components` entry,
+- Flatcar version and image variant MUST BE specified as `.spec.components` entries,
```

Comment:
Is this purely informational or is it used to decide what VM image to use? As we also bake Teleport agent into the node image does that also need to be included here?

Reply:
It decides which image to use.

The idea is to get all versions of all apps / components from the Release, so:
- Kubernetes version
- OS version
  - info needed for the whole image name, including variant number that changes when we get e.g. new Teleport version baked in
  - alternatively we can also put the entire image name so it's more flexible (e.g. cluster-aws currently allows to override the image name with a free-form string, or it looks up image according to defined format, which is what we usually use and what I also did here)
- app versions for apps that are deployed as App CRs
- app versions for apps that are deployed as HelmRelease CRs